src/java/main/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1054,6 +1054,8 @@ private void sendPing() {
         private boolean saslLoginFailed = false;
 
         private void startConnect() throws IOException {
+            // initializing it for new connection
+            saslLoginFailed = false;
```

Comment:
Successfull Login is not necessary to successfully connect to ZK Server.  ZK server may allow connection without client login being successful, based on configuration. So it is better to set the flag at the start of the new connection.  
This change has impact on tunnelAuthInProgress, in this method also same thing is expeted. In this there is a sceanrio where zooKeeperSaslClient can be null